### DISCUSSION TITLE
Why was the support for `<--` comments removed?

### DISCUSSION THREAD

**Discussion by ovflowd** [mdx-js/mdx#2219] [Q&A] [answered]
I understand that MDX wants to encourage you to use JSX comments. But when you have use cases like legacy Apps with hundreds and hundreds of block comments `<--` it makes the migration extremely painful. (I might be missing something here)

I'm just genuinely curious about the historical decision process behind it and why there are no "compatibility" plugins/options for MDX.

**Comment by wooorm** [comment, marked as answer]
> extremely painful

Use a regex to replace `<!--` with `{/*` and `-->` with `*/}`, or a plugin, or use the `.md` extension if you want markdown.

> randomly decide

No, it was a very careful decision to make what is markdown and what is JS(X) unambiguous. If you want to migrate from markdown, you don’t need to migrate, use an `.md` extension.

> genuinely curious

- https://github.com/mdx-js/mdx/issues/1042
- https://github.com/wooorm/xdm/issues/44
- https://github.com/wooorm/xdm/issues/61

> "compatibility" plugins/options

a) there is a plugin.
b) because MDX before v2 passed everything that looked like XML through as JS(X) to Babel/Webpack. We can’t codemod that. We can’t support that. From v2 on, we have an actual AST, which can be transformed, so we can do that.

**Reply by ovflowd** [reply]
Thanks for the fast reply! I've added some notes on a separate comment, but I appreciate the quick and thorough reply :)

**Comment by ovflowd** [comment]
> Use a regex to replace with/}, or a plugin, or use the .md` extension if you want markdown.

This is not a feasible option when you have hundreds of files. I cannot feasibly ask the other collaborators to review each one of the 700+ file changes and guarantee them that these changes are safe. (Assuming I'm going the regex route)

> No, it was a very careful decision to make what is markdown and what is JS(X) unambiguous. If you want to migrate from markdown, you don’t need to migrate, use an .md extension.

If I import a plain Markdown file inside MDX, it will be treated as MDX. So, there's an issue here, not sure if it's a bug or per design. But that's minor, honestly.

> a) there is a plugin.

Which plugin? And what does that plugin does?

**Reply by ovflowd** [reply]
Having Kent creating [this](https://github.com/wooorm/xdm/issues/61) issue doesn't really inspire much confidence regarding if this was a sustainable solution.

Sure, MDX is MDX, and you want to be strict on what you support. But there should be a way to go around that, even more for such breaking changes that MDXv2 causes.

This is, for example, a huge issue when we (Node.js) want to consider the adoption of MDX.

**Reply by ovflowd** [reply]
I am aware I definitely lack context here and that I might sound harsh; just someone trying to quickly see what options are available and how I might get around this issue on a long-term plan.

Not criticizing the decision made but wondering how it got there and the current options to get around or safely be "compatible" without necessarily regexing hundreds of files. 😅

**Reply by wooorm** [reply]
You do lack a context, and seem to reply fast, while not reading the issues: the plugin is shown in the first linked issue.

> Having Kent creating https://github.com/wooorm/xdm/issues/61 issue doesn't really inspire much confidence regarding if this was a sustainable solution.

You seem to not have read the threads, found a famous name, and concluded from that that it was bad.

That is not the case, Kent was the first person to build something very popular (`mdx-bundler`) on MDX 2 (back then, xdm) and ship it to a lot of people. He here, 18 months ago, asks about the new style of comments in combination with dev tooling (prettier), which didn‘t support them yet. Then we added support to these comments in Prettier.

> If I import a plain Markdown file inside MDX, it will be treated as MDX

This seems your issue. Please ask about this (and read the docs: https://mdxjs.com/docs/what-is-mdx/#mdx-syntax). This is also mentioned in the issues I linked.

> I cannot feasibly ask the other collaborators

Yes you can.

> But there should be a way to go around that, even more for such breaking changes that MDXv2 causes.

There is (plugin, regex) but you shouldn’t: use standard MDX or markdown.

> This is, for example, a huge issue when we (Node.js) want to consider the adoption of MDX.

I don‘t think it is. You are asking about XY problems. You are asking about comments but I believe you want to support markdown.

> I am aware I definitely lack context here and that I might sound harsh; just someone trying to quickly see what options are available and how I might get around this issue on a long-term plan.

You seem to be aware of how open source works. I recommend not being “quick” about it. Spending the time up front to explain the actual situation. Zooming out.

> Not criticizing the decision made but wondering how it got there and the current options to get around or safely be "compatible" without necessarily regexing hundreds of files. 😅

Yes, you are criticizing. Which you can do. But it’s a good decision. There is no need to get around things. Use `.md` files if you want markdown.

**Reply by ovflowd** [reply]
> You do lack a context, and seem to reply fast, while not reading the issues: the plugin is shown in the first linked issue.

Yes, I am aware I rushed through everything, as I was in the middle of 3 different things and ADHD didn't really make it easer. I apologise for that 😅 

> You seem to not have read the threads, found a famous name, and concluded from that that it was bad.

True, and that's on me, my bad!

> That is not the case, Kent was the first person to build something very popular (mdx-bundler) on MDX 2 (back then, xdm) and ship it to a lot of people. He here, 18 months ago, asks about the new style of comments in combination with dev tooling (prettier), which didn‘t support them yet. Then we added support to these comments in Prettier.

I have no idea what mdx-bundler and xdm is, but thanks for the explanation. Will give a more throughout read afterwards :)

> There is (plugin, regex) but you shouldn’t: use standard MDX or markdown.

I think that's the issue I'm having, sometimes MDX (maybe intentionally or not sure how) is interpreting plain MD as MDX. E.g.: complaining about `div class=""` on a md file 🤔 

> You seem to be aware of how open source works. I recommend not being “quick” about it. Spending the time up front to explain the actual situation. Zooming out.

Yup, sadly I didn't have the time of zooming out and reading carefully, I know that's on me, I still appreciate you going through my comments :)

> Yes, you are criticizing. Which you can do. But it’s a good decision. There is no need to get around things. Use .md files if you want markdown.

Sorry if I sounded like criticizing, I genuinely wanted to understand the "why", if that's criticizing then, well, I learned a new thing today. (language barriers are fantastic).

Thank you again for the replies!

**Reply by wooorm** [reply]
> Yes, I am aware I rushed through everything, as I was in the middle of 3 different things and ADHD didn't really make it easer. I apologise for that 😅

💙

> I have no idea what mdx-bundler and xdm is, but thanks for the explanation. Will give a more throughout read afterwards :)

xdm was to MDX what IO.js was to Node.js!

> complaining

Can you expand when MDX is complaining?
The `class` vs `className` difference is a React thing, that the MDX projects don’t touch.

> I genuinely wanted to understand the "why", if that's criticizing then, well

> language barriers are fantastic

Quite! I’m not a native English speaker myself. 
Trying to understand something isn‘t criticising. And, criticising isn’t a problem.

You used language like “extremely painful” and “randomly decide”, to which there are no acceptable answers that can be given.
Also words like “genuinely”, shouldn‘t be needed when someone is genuine, and language such as “Not xyz **but**”, you can do a google/duckduckgo search on `everything before the but` for more on that!

**Reply by ovflowd** [reply]
> xdm was to MDX what IO.js was to Node.js!

Gotcha!

> Can you expand when MDX is complaining?
The class vs className difference is a React thing, that the MDX projects don’t touch.

Interesting, so basically we have plain HTML inside a regular `md` file https://github.com/ovflowd/nodejs.org/blob/feat/nextjs-infrastructure/pages/en/docs/index.mdx?plain=1#L24 (Note that I already renamed this file to MDX and fixed the error here), but I tried to search dep on previous issues but didn't find a way to make MDX to ignore "JSX'ing" of plain HTML.

> You used language like “extremely painful” and “randomly decide”, to which there are no acceptable answers that can be given.

To be frank, I've had some roadblocks with MDX in a few repositories before, let's say I had an emotional rush from those previous encounters. It was inappropriate and I apologise. (as it had nothing to do with this discussion).

**Reply by wooorm** [reply]
> Interesting, so basically we have plain HTML inside a regular md file [ovflowd/nodejs.org@feat/nextjs-infrastructure/pages/en/docs/index.mdx?plain=1#L24](https://github.com/ovflowd/nodejs.org/blob/feat/nextjs-infrastructure/pages/en/docs/index.mdx?plain=1#L24) (Note that I already renamed this file to MDX and fixed the error here), but I tried to search dep on previous issues but didn't find a way to make MDX to ignore "JSX'ing" of plain HTML.

I don’t understand. You show an MDX file, which uses JSX, with `className`. I need to know what input gave what actual output and what you expected instead. Perhaps your issue is already solved?
I also don’t understand what “search dep” is.
Re `JSX’ing of plain HTML`, I am also not sure what you mean. What is plain HTML? And there is no HTML in MDX, there is only JSX.

**Reply by ovflowd** [reply]
> I don’t understand. You show an MDX file, which uses JSX, with className. 

I mentioned that this file is the final result of me fixing the error 😅 (It was a `.md` file before and with `class`). It was giving the generic React error to use `className` instead of `class`

> I also don’t understand what “search dep” is.

I meant to write "search deep" (aka googled previous issues)"

> Re JSX’ing of plain HTML, I am also not sure what you mean. What is plain HTML? And there is no HTML in MDX, there is only JSX.

MDX should not parse (IMHO) HTML tags within Markdown files. `<div something something>` should be untouched by MDX, I guess?

**Reply by wooorm** [reply]
Thank you for sponsoring my work btw!! ✨ Appreciated!

> MDX should not parse (IMHO) HTML tags within Markdown files. `<div something something>` should be untouched by MDX, I guess?

I think this is very important: these are not HTML tags, they are JSX components. MDX is not an HTML parser. MDX is a markdown minus XML-like things plus an actual JS(X) parser. That is a JSX tag, equivalent to:

```
React.createElement("div", {
  something: true,
  something: true
});
```

…assuming React.

**Reply by ovflowd** [reply]
> Thank you for sponsoring my work btw!! ✨ Appreciated!

Enough for a coffee to relax after this discussion xD (Jokes aside, even tho this heated discussion, I definitely appreciate your work with unifed/remark/mdx)

> I think this is very important: these are not HTML tags, they are JSX components. MDX is not an HTML parser. MDX is a markdown minus XML-like things plus an actual JS(X) parser. That is a JSX tag, equivalent to:

See, I think here's where my confusion comes from. I would expect `.md` files to behave as regular Markdown files and have Remark responsible for rendering them to HTML. And then MDX would actually have the JSX stuff.

Thanks for clarifying!

**Reply by wooorm** [reply]
Oh, but now I am talking about the MDX format, with that explicit extension. If the extension is .md, then it is HTML, not JSX. If the extension is .mdx, then it is JSX, not HTML. If that isn't the case, I'd like to see a small reproduction!

**Reply by ovflowd** [reply]
> Oh, but now I am talking about the MDX format, with that explicit extension. If the extension is .md, then it is HTML, not JSX. If the extension is .mdx, then it is JSX, not HTML. If that isn't the case, I'd like to see a small reproduction!

Yup, that's what I was trying to say, the file mentioned (here https://github.com/ovflowd/nodejs.org/blob/feat/nextjs-infrastructure/pages/en/docs/index.mdx?plain=1#L24) before the current commits was on `.md` with the `<div class="highligh-box" />` and React was complaining about the `class` to `className` transition. Hence why I was confused 😅 

Not sure if it is a bug.

**Reply by ovflowd** [reply]
cc @shuding (maybe you know why plain HTML tags are being handled as JSX within the MDX installation of Nextra?

(Nextra is a flavour of Next.js, https://nextra.site)

**Reply by ovflowd** [reply]
Maybe the MDX setup that Nextra is doing is incorrec.t @wooorm is there something we need to pass to the processor options from `mdx-js` to treat .md` as regular MD?

**Reply by remcohaszing** [reply]
This line processes the source from string. https://github.com/shuding/nextra/blob/238dea3881165df3244c2b3a23944ac438b3721d/packages/nextra/src/compile.ts#L94

MDX infers MDX/markdown mode based on the file path. You should pass a vfile compatible object instead

```js
{
  value: source,
  path: filePath
}
```

**Reply by ovflowd** [reply]
@shuding ☝️

**Reply by ovflowd** [reply]
Btw @remcohaszing thank you for going deep and finding this!

**Reply by remcohaszing** [reply]
Sorry, my bad! `compile()` does this internally, then passes the format option to `createProcessor()`. So you’ll need to do something like this instead ([`resolveFileAndOptions()`](https://github.com/mdx-js/mdx/blob/main/packages/mdx/lib/util/resolve-file-and-options.js#L18-L35)):

```ts
createProcessor({
  format: filePath.toLowerCase().endsWith('.md') ? 'markdown' : 'mdx'
})
```

Still, it’s a good idea to pass `path` to `process`. External plugins may use this.
